Public/Src/Tools/SymbolDaemon/SymbolDaemon.cs

```diff
@@ -218,11 +229,117 @@ private static Client CreateClient(string serverMoniker, IClientConfig config)
+            Contract.Assert(!hashes.Any(hash => hash.HashType != HashType.Vso0), "Unsupported hash type");
+
+            var indexer = new SymbolIndexer(SymbolAppTraceSource.SingleInstance);
+            var symbolsMetadata = new Dictionary<ContentHash, HashSet<DebugEntryData>>();
+
+            for (int i = 0; i < files.Length; i++)
+            {
+                HashSet<DebugEntryData> symbols;
+                if (!symbolsMetadata.TryGetValue(hashes[i], out symbols))
+                {
+                    symbols = new HashSet<DebugEntryData>(DebugEntryDataComparer.Instance);
+                    symbolsMetadata.Add(hashes[i], symbols);
+                }               
+
+                // Index the file. It might not contain any symbol data. In this case, we will have an empty set.
+                symbols.UnionWith(indexer.GetDebugEntries(new System.IO.FileInfo(files[i]), calculateBlobId: false));               
+            }
+            
+            SerializeSymbolsMetadata(symbolsMetadata, outputFile);
+        }
+
+        /// <summary>
+        /// Serializes DebugEntryData's into a file.
+        /// </summary>        
+        public static void SerializeSymbolsMetadata(Dictionary<ContentHash, HashSet<DebugEntryData>> symbolsMetadata, string outputFile)
+        {
+            /*                
+                <number of hashes>
+                for each hash:
+                    <hash>
+                    <number of debug entries>
+                    [<debug entry>]
+             */
+             
+            using (var writer = new StreamWriter(outputFile))
+            {                
+                writer.WriteLine(symbolsMetadata.Count);
+
+                foreach (var kvp in symbolsMetadata)
+                {
+                    writer.WriteLine(kvp.Key.Serialize());
+                    writer.WriteLine(kvp.Value.Count);
+
+                    foreach (var debugEntry in kvp.Value)
+                    {
+                        writer.WriteLine(JsonConvert.SerializeObject(debugEntry));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Deserializes DebugEntryData's from a file.
+        /// </summary>
+        public static Dictionary<ContentHash, HashSet<DebugEntryData>> DeserializeSymbolsMetadata(string fileName)
+        {
+            var result = new Dictionary<ContentHash, HashSet<DebugEntryData>>();
+            using (var reader = new StreamReader(fileName))
+            {
+                int hashCount = int.Parse(reader.ReadLine());
+
+                for (int i = 0; i < hashCount; i++)
+                {
+                    ContentHash.TryParse(reader.ReadLine(), out var hash);
+
+                    var symbols = new HashSet<DebugEntryData>(DebugEntryDataComparer.Instance);
+                    result.Add(hash, symbols);
+
+                    int debugEntryCount = int.Parse(reader.ReadLine());

```

Comment:
If kvp.Value is serialized as an object, you can deserialized the written Json Object into an array and get the count and entries easily.